### ISSUE TITLE
Inference on trained model to generate unsupervised segmentation maps on any dataset

### ISSUE THREAD

**Issue by ramprs** [xu-ji/IIC#89]
Hi, could you tell me how I can just run inference given a folder containing raw images to create segmentation maps? Thanks in advance

**Comment by xu-ji** [comment]
Download the trained model of your choice. Write a script to load the model and run your images through, get the predictions as the output of the network, then use e.g. matplotlib to render the output. 

You probably will need to process your images in some way to fit the same format I used (e.g. sobel filter, resize).

**Comment by OmarHedeya95** [comment]
Would someone please like to elaborate more how to do this?

**Comment by OmarHedeya95** [comment]
If someone is stuck I did it the following way, I do not know if it is wrong or not, but I got an image somehow resembling my input and looks segmented. If you notice a mistake please let me know!

config_in = open("./models/555/config.pickle", "rb")
config = pickle.load(config_in)
net = archs.__dict__[config.arch](config)
net.load_state_dict(torch.load("./models/555//best_net.pytorch"))


image = Image.open('./Horse.jpg')
greyscale = custom_greyscale_to_tensor(include_rgb=True)

im = greyscale(image)
im.unsqueeze_(0)

print(im.shape)

im = im.float()
im = im.cuda()

print(im.shape)

im = im.cpu()

net.eval()

output = net(im)
output_image = output[0]

#Detach to stop the tensor from needing a gradient!
output_image = output_image.detach()
#To remove the first dimension(batch dimension)
output_image = output_image.squeeze()

#To turn the tensor image into a format where channel is the last dimension
plt.imshow(output_image.permute(1, 2, 0))
plt.show()

**Comment by ramprs** [comment]
Hi @xu-ji, after getting the output from the network, do you recommend using any additional postprocessing such as clustering? Here is the network output I am getting for [this](https://images.cocodataset.org/val2017/000000173830.jpg) image. Is this the expected output? 
![output](https://user-images.githubusercontent.com/19614537/133834601-588e160c-01fe-4e41-a95f-3c5069ac37f8.png)
Thanks in advance

**Comment by xu-ji** [comment]
Post-processing can help of course but it should not be needed. We did not do any post-processing in the paper. How many clusters are you using? Consider increasing it so that there is better discrimination of the background.

**Comment by ramprs** [comment]
Thank you for your reply @xu-ji . May I know how to set the number of clusters? I am attaching the code I am using. Could you tell me where I need to make this change or if there is something wrong in the way I am using the models/functions? 

Thank you so much in advance. 

```

from code.utils.cluster.transforms import sobel_process, custom_greyscale_numpy`

config_in = open("./models/555/config.pickle", "rb")
config = pickle.load(config_in)
net = archs.__dict__[config.arch](config)
net.load_state_dict(torch.load("./models/555/best_net.pytorch"))


# load image from memory
image = cv2.imread('./coco/images/val2017/000000173830.jpg')

im = custom_greyscale_numpy(image, include_rgb=True)
im = im.astype(np.float32) / 255.
im = torch.from_numpy(im).permute(2, 0, 1).unsqueeze(0)
im_sobel = sobel_process(im, True)
im = torch.autograd.Variable(im_sobel)

net.eval()

output = net(im)
output_image = output[0].detach().squeeze()

#To turn the tensor image into a format where channel is the last dimension
plt.imshow(output_image.permute(1, 2, 0))
plt.show()
```

**Comment by ShiAngWang** [comment]
> Thank you for your reply @xu-ji . May I know how to set the number of clusters? I am attaching the code I am using. Could you tell me where I need to make this change or if there is something wrong in the way I am using the models/functions?
> 
> Thank you so much in advance.
> 
> ```
> 
> from code.utils.cluster.transforms import sobel_process, custom_greyscale_numpy`
> 
> config_in = open("./models/555/config.pickle", "rb")
> config = pickle.load(config_in)
> net = archs.__dict__[config.arch](config)
> net.load_state_dict(torch.load("./models/555/best_net.pytorch"))
> 
> 
> # load image from memory
> image = cv2.imread('./coco/images/val2017/000000173830.jpg')
> 
> im = custom_greyscale_numpy(image, include_rgb=True)
> im = im.astype(np.float32) / 255.
> im = torch.from_numpy(im).permute(2, 0, 1).unsqueeze(0)
> im_sobel = sobel_process(im, True)
> im = torch.autograd.Variable(im_sobel)
> 
> net.eval()
> 
> output = net(im)
> output_image = output[0].detach().squeeze()
> 
> #To turn the tensor image into a format where channel is the last dimension
> plt.imshow(output_image.permute(1, 2, 0))
> plt.show()
> ```

Hi， have you any idea about how should I resize the segmentation result image to the original image size, because the original image is resized to 1128x128, the result is 128x128. 
And I have some not very satisfying segmentation results just like the following on CoCo-Stuff-3, do you have more ideas about using the author's model to output the segmentation result?

original:
![image](https://user-images.githubusercontent.com/26656883/140498936-1fd35e97-58cb-4a15-93b3-8aa2e1d20342.png)

result:
![image](https://user-images.githubusercontent.com/26656883/140498951-092afec0-5856-437f-a184-efe5d087d028.png)
